design/20220118.server-side-apply.md

```diff
@@ -0,0 +1,167 @@
+string `cert-manager`:
+
+```text
+cert-manager-<cert-manager component>
+```
+
+e.g.
+
+```text
+cert-manager-certificates-issuing
+```
+
+### User Agent
+
+The user agent is inspired by the [client-go default user
+agent](https://github.com/kubernetes/client-go/blob/664b1a6c8ce9c92ce65bef3f9833b402449c98d2/rest/config.go#L499),
+and is defined with the following:
+
+```text
+<component field manager>/<cert-manager version> (<Operating System>/<Architecture>) cert-manager/<git commit>
+```
+
+e.g:
+
+```text
+cert-manager-certificates-issuing/v1.7.0 (Linux/amd64) cert-manager/0b686b8f38c8c7442744c9224d18e780ee7f244a
+```
+
+## Implementation Consideration
+
+All CREATE and UPDATE operations in cert-manager controllers are to be replaced
+with APPLY. For this change, controllers will need to be modified so that for
+these calls, _only_ the fields in which they manage should be included in the
+PATCH API call. This means each controller will need to define exactly which
+fields they are concerned with.
+
+Some fields, such as the Certificate Issuing Condition are managed by more than
+one controller (issuing and trigger Certificate controllers), and as such, will
+need to make use of the `force` option in their API calls. This option tells the
+API server to revoke management of that field from the previous owner, overwrite
+the field, and change owner ship to the new client.
+
+The [fake client-go client](https://github.com/kubernetes/client-go/issues/970)
+does not support the Apply patch call for mocking API calls and events. This
+means that significant controller unit-testing will either need to moved to
+testing against a real API server as integration tests, the controller
+[test framework must add custom support for Apply](https://github.com/jetstack/cert-manager/blob/master/pkg/controller/test/context_builder.go),
+or a new testing framework should be developed.
+
+Some API fields will need to have some metadata updated to function better under
+Server Side Apply. One such example is adding `x-kubernetes-list-type=map` and
+`x-kubernetes-list-map-keys=Type` to the [Certificates Status Condition
+slice](https://github.com/jetstack/cert-manager/blob/0ca1ce9a6a1d7c311afd4b3e786975759249132a/pkg/apis/certmanager/v1/types_certificate.go#L385),
+so that controllers are able to apply distinct condition types, without
+conflicting with other controller conditions (i.e. the Ready and Issuing
+conditions).
+
+## Feature Gate
+
+Placing the APPLY functionality behind a feature gate should be considered.
+Placing this functionality behind a feature gate would allow the cert-manager
+authors gain confidence about its correctness, and ensure there are no
+regressions in the stability of controller reconciliation.
+
+To reach graduation, the cert-manager authors should consider the server side
+apply implementation for cert-manager to be safe for end users. Graduation
+should ideally be done over no more than one or two releases.
```

Comment:
How can we validate it is safe? Are there metrics we can use? (e.g. "we expect to see a decrease in the number of resource version conflicts"?

Will we directly solicit feedback from users?

> Graduation should ideally be done over no more than one or two releases.

I think it's great to timebox this, however if we *don't* have confidence, we should not proceed with graduation. (and I think it's worth noting that here too)

Reply:
Yep, this is something we should define. I'll add a note about soliciting user feedback, but we should keep thinking of other ideas (if there is anything else we can do).

Indeed, this was meant as a rfc `SHOULD` rather than `MUST`. I'll make that more clear